reflex/state.py

```diff
@@ -1078,7 +1266,75 @@ def set_state(self, token: str, state: State):
         """
         if self.redis is None:
             return
-        self.redis.set(token, cloudpickle.dumps(state), ex=self.token_expiration)
+        await self.redis.set(token, cloudpickle.dumps(state), ex=self.token_expiration)
+
+    @contextlib.asynccontextmanager
+    async def _redis_lock(self, token: str):
+        """Obtain a redis lock for a token.
+
+        Args:
+            token: The token to obtain a lock for.
+
+        Yields:
+            None when the lock is held
+
+        Raises:
+            RuntimeError: If redis is not configured.
+        """
+        if self.redis is None:
+            raise RuntimeError("Redis is not configured")
+        lock_key = f"{token}_lock".encode()
+        # TODO: set key expiry
+        state_is_locked = await self.redis.setnx(lock_key, 1)
+        while not state_is_locked:
```

Comment:
considered; now we raise a LockExpiredError if the mutator holds the lock for too long and we reject their changes so that other waiters don't get corrupted. The exception message suggests to the developer how to increase the lock_expiration timeout or use background tasks.